src/core/qgsvectorlayerjoinbuffer.cpp

```diff
@@ -541,6 +541,7 @@ bool QgsVectorLayerJoinBuffer::addFeatures( QgsFeatureList &features, QgsFeature
         const QString filter = QgsExpression::createFieldEqualityExpression( info.joinFieldName(), idFieldValue.toString() );
 
         QgsFeatureRequest request;
+        request.setFlags( QgsFeatureRequest::NoGeometry );
         request.setFilterExpression( filter );
```

Comment:
you should also setSubsetOfAttributes( QgsAttributeList() ) here - all the iterators have logic and tests to ensure that using a filter expression will always add the required attributes for the expression to be fetched.